File: 62.不同路径.js
/*
 * @lc app=leetcode.cn id=62 lang=javascript
 *
 * [62] 不同路径
 * 62/62 cases passed (64 ms)
 * Your runtime beats 87.41  % of javascript submissions
 * Your memory usage beats 37.11 % of javascript submissions (34.9 MB)
 */

// @lc code=start
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */

var uniquePaths = function(m, n) {
let map = [];
  for (let i = 0; i < m; i++) {
    map[i] = [];
    for (let j = 0; j < n; j++) {
        if(i == 0 || j == 0){
            map[i][j] = 1;
        } else {
            map[i][j] = map[i - 1][j] + map[i][j - 1];
        }
    }
  }
  return map[m-1][n-1];
};
// @lc code=end

console.log(uniquePaths(3, 2));
console.log("end");

File: 62.不同路径.java
/*
 * @lc app=leetcode.cn id=62 lang=java
 *
 * [62] 不同路径
 */
class Main {
    public static void main(String[] args) {
        // Create a new Solution62 instance
        Solution62 Solution62 = new Solution62();
        // Create a test case
        // String testCase = "()[]{}";
        // Get the answer
        // boolean answer = Solution62.uniquePaths(3,2);
        // Print the answer
        System.out.println(Solution62.uniquePaths(3,2));
    }
}

// @lc code=start
class Solution62 {
    // HashMap<Integer, Integer> map = new HashMap<>();
    // public  boolean divisorGame(int N) {
    //     if ( map.get(N) != null && map.get(N) == true) {
    //         return map.get(N);
    //     }
    //     for (int i = 1; i < N; i++) {
    //         if (N % i == 0) {
    //             boolean res = divisorGame(N - i);
    //             if (map.get(N-i) == null) {
    //                  map.put(N-i, res);
    //             }
    //             return !res;
    //         }
    //     }
    //    return false;
    // }

    
    int[][] map = new int[101][101]; 
    public int uniquePaths(int m, int n) {
        System.out.println("first1="+map[m][n]);
        if(map[m][n] !=0){
            return map[m][n];
        } else {
            if(m == 1 || n == 1){
                map[m][n] = 1;
            }
            map[m][n] = uniquePaths(m-1,n) + uniquePaths(m,n-1);
        }
        return map[m][n];
        // return 1;
    }
}
// @lc code=end



// class Solution62 {

//     ...

//     public boolean isValid(String s) {

//         ...

//         return answer;
//     }
// }